components/aws/sagemaker/tests/unit_tests/tests/test_hpo.py

```diff
@@ -60,4 +64,286 @@ def test_empty_string(self):
+                               'EnableManagedSpotTraining': False}
+    )
+
+    self.assertEqual(response, 'test-job')
+  
+
+  def test_sagemaker_exception_in_create_hyperparameter_tuning_job(self):
+    mock_client = MagicMock()
+    mock_exception = ClientError({"Error": {"Message": "SageMaker broke"}}, "create_hyperparameter_tuning_job")
+    mock_client.create_hyper_parameter_tuning_job.side_effect = mock_exception
+    mock_args = self.parser.parse_args(required_args)
+
+    with self.assertRaises(Exception):
+       _utils.create_hyperparameter_tuning_job(mock_client, vars(mock_args))
+ 
+  def test_wait_for_hyperparameter_tuning_job(self):
+    mock_client = MagicMock()
+    mock_client.describe_hyper_parameter_tuning_job.side_effect = [
+      {"HyperParameterTuningJobStatus": "InProgress"},
+      {"HyperParameterTuningJobStatus": "Completed"},
+      {"HyperParameterTuningJobStatus": "Should not be called"}
+    ]
+
+    _utils.wait_for_hyperparameter_training_job(mock_client,'hyperparameter-tuning-job',0)
+    self.assertEqual(mock_client.describe_hyper_parameter_tuning_job.call_count, 2)
+  
+  def test_wait_for_failed_job(self):
+    mock_client = MagicMock()
+    mock_client.describe_hyper_parameter_tuning_job.side_effect = [
+      {"HyperParameterTuningJobStatus": "InProgress"},
+      {"HyperParameterTuningJobStatus": "Failed", "FailureReason": "Something broke lol"},
+      {"HyperParameterTuningJobStatus": "Should not be called"}
+    ]
+
+    with self.assertRaises(Exception):
+      _utils.wait_for_hyperparameter_training_job(mock_client, 'training-job', 0)
+
+    self.assertEqual(mock_client.describe_hyper_parameter_tuning_job.call_count,2)
+
+  def test_get_image_from_algorithm_job(self):
+    mock_client = MagicMock()
+    mock_client.describe_hyper_parameter_tuning_job.return_value = {"TrainingJobDefinition": {"AlgorithmSpecification": {"AlgorithmName": "my-algorithm"}}}
+    mock_client.describe_algorithm.return_value = {"TrainingSpecification": {"TrainingImage": "training-image-url"}}
+
+    self.assertEqual(_utils.get_image_from_job(mock_client, 'training-job'), "training-image-url")
+
+  def test_reasonable_required_args(self):
+    response = _utils.create_hyperparameter_tuning_job_request(vars(self.parser.parse_args(required_args)))
+
+    # Ensure all of the optional arguments have reasonable default values
+    self.assertFalse(response['TrainingJobDefinition']['EnableManagedSpotTraining'])
+    self.assertDictEqual(response['TrainingJobDefinition']['StaticHyperParameters'], {})
+    self.assertNotIn('VpcConfig', response['TrainingJobDefinition'])
+    self.assertNotIn('MetricDefinitions', response['TrainingJobDefinition'])
+    self.assertEqual(response['Tags'], [])
+    self.assertEqual(response['TrainingJobDefinition']['AlgorithmSpecification']['TrainingInputMode'], 'File')
+    self.assertEqual(response['TrainingJobDefinition']['OutputDataConfig']['S3OutputPath'], 'test-output-location')
+
```

Comment:
Can we add some more good defaults into this test? This is just sanity checking that the default arguments produce proper hyperparameters and solid default outputs/inputs.

Reply:
I think its getting tested in `test_create_hyperparameter_tuning_job` with all default value that`create_hyper_parameter_tuning_job` uses.

Reply:
Sure yeah that sounds fine.